compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1220,28 +1229,102 @@ impl<'a> Parser<'a> {
+                } else {
+                    self.sess.emit_err(errors::StaticWithGenerics { span: generics.span });
+                }
+            }
+            generics
+        } else {
+            Generics::default()
+        };
+
         // Parse the type of a `const` or `static mut?` item.
         // That is, the `":" $ty` fragment.
-        let ty = match (self.eat(&token::Colon), self.check(&token::Eq) | self.check(&token::Semi))
-        {
-            // If there wasn't a `:` or the colon was followed by a `=` or `;` recover a missing type.
+        // FIXME: This could maybe benefit from `.may_recover()`?
+        let ty = match (
+            self.eat(&token::Colon),
+            self.check(&token::Eq) | self.check(&token::Semi) | self.check_keyword(kw::Where),
+        ) {
+            // If there wasn't a `:` or the colon was followed by a `=`, `;` or `where`, recover a missing type.
             (true, false) => self.parse_ty()?,
             (colon, _) => self.recover_missing_const_type(colon, m),
         };
 
+        let before_where_clause =
+            if self.may_recover() { self.parse_where_clause()? } else { WhereClause::default() };
+
         let expr = if self.eat(&token::Eq) { Some(self.parse_expr()?) } else { None };
+
+        let after_where_clause = if m.is_none() || self.may_recover() {
+            self.parse_where_clause()?
+        } else {
+            WhereClause::default()
+        };
+
+        if before_where_clause.has_where_token && m.is_none() && let Some(expr) = &expr {
+            self.sess.emit_err(errors::WhereClauseBeforeConstBody {
+                span: before_where_clause.span,
+                name: id.span,
+                body: expr.span,
+                sugg: if !after_where_clause.has_where_token {
+                    self.sess.source_map().span_to_snippet(expr.span).ok().map(|body| {
+                        errors::WhereClauseBeforeConstBodySugg {
+                            left: before_where_clause.span.shrink_to_lo(),
+                            snippet: body,
+                            right: before_where_clause.span.shrink_to_hi().to(expr.span),
+                        }
+                    })
+                } else {
+                    // FIXME(generic_const_items): Provide a structured suggestion to merge the first
+                    // where-clause into the second one.
+                    None
+                },
+            });
+        }
+
+        let mut predicates = before_where_clause.predicates;
+        predicates.extend(after_where_clause.predicates);
+        let where_clause = WhereClause {
+            has_where_token: before_where_clause.has_where_token
+                || after_where_clause.has_where_token,
+            predicates,
+            span: if after_where_clause.has_where_token {
+                after_where_clause.span
+            } else {
+                before_where_clause.span
+            },
+        };
```

Comment:
> a well-formed constant can have predicates in before_where_clause if may_recover() and expr.is_none().

You are absolutely right, checking `=` beforehand might just simplify things :+1: I thought about that too at some point but I left it at mirroring the type alias parser which isn't really necessary actually.

Reply:
Update: `parse_const_item` stills contains two calls to `parse_where_clause` and the predicate merging business. I don't see how this can be simplified while meeting all requirements. I'd like to (at least) provide a good diagnostic for where-clauses before `=` as well as for WC-eq-body-WC (and I'd prefer to also nicely recover from both).

> by checking if we can eat a `=` before trying to parse a `where`

Seeing as this is equivalent to getting rid of the first `parse_where_clause` (given that we otherwise already `check` for the presence of `=` to decide if we want to parse a body or not), this doesn't allow for error recovery on WC-eq-body. I hope I'm not missing something obvious here :eyes:.